src/data_types/tensor.hpp

```diff
@@ -0,0 +1,410 @@
+#include "tensor_index_tools.hpp"
+#include "vector_index_tools.hpp"
+
+namespace detail {
+
+template <class ValidIndexSet>
+struct ToCoord;
+
+template <class... Dims>
+struct ToCoord<VectorIndexSet<Dims...>>
+{
+    using type = Coord<Dims...>;
+};
+
+template <class ValidIndexSet>
+using to_coord_t = typename ToCoord<ValidIndexSet>::type;
+
+} // namespace detail
+
+/**
+ * @brief A class representing a Tensor.
+ * @tparam ElementType The type of the elements of the tensor (usually double/complex).
+ * @tparam ValidIndexSet The indices that can be used along each dimension of the tensor.
+ */
+template <class ElementType, class... ValidIndexSet>
+class Tensor
+{
+    static_assert(std::conjunction_v<tensor_tools::is_vector_index_set<ValidIndexSet>...>);
+    static_assert(std::conjunction_v<std::disjunction<
+                          tensor_tools::is_covariant_vector_index_set<ValidIndexSet>,
+                          tensor_tools::is_contravariant_vector_index_set<ValidIndexSet>>...>);
+
+    using AllIndexSets = ddc::detail::TypeSeq<ValidIndexSet...>;
+
+public:
+    /// The TensorIndexSet describing the possible indices.
+    using index_set = tensor_tools::TensorIndexSet<ValidIndexSet...>;
+
+private:
+    static constexpr std::size_t s_n_elements = (ddc::type_seq_size_v<ValidIndexSet> * ...);
+    std::array<ElementType, s_n_elements> m_data;
+
+public:
+    /// The type of the elements of the tensor.
+    using element_type = ElementType;
+
+    /**
+     * @brief The rank of the tensor.
+     * This is equivalent to the number of indices required to
+     * access an element of the tensor.
+     *
+     * @return The rank of the tensor.
+     */
+    KOKKOS_FUNCTION static constexpr std::size_t rank()
+    {
+        return sizeof...(ValidIndexSet);
+    }
+
+    /**
+     * @brief The size of the tensor.
+     * This is the number of elements in the tensor.
+     *
+     * @return The number of elements in the tensor.
+     */
+    KOKKOS_FUNCTION static constexpr std::size_t size()
+    {
+        return s_n_elements;
+    }
+
+private:
+    static_assert(rank() > 0);
+
+public:
+    /**
+     * @brief Construct an uninitialised tensor object.
+     */
+    explicit KOKKOS_FUNCTION Tensor() {}
+
+    /**
+     * @brief Construct a tensor object initialised with a value.
+     * @param fill_value The value with which the tensor should be filled.
+     */
+    explicit KOKKOS_FUNCTION Tensor(ElementType fill_value)
+    {
+        m_data.fill(fill_value);
+    }
+
+    /**
+     * @brief Construct a 1D tensor object by providing the elements that should
+     * be saved in the vector.
+     *
+     * @param elements The elements of the tensor.
+     */
+    template <
+            class... Params,
+            class = std::enable_if_t<(std::is_convertible_v<Params, ElementType> && ...)>,
+            class = std::enable_if_t<sizeof...(Params) == size()>>
+    KOKKOS_FUNCTION Tensor(Params... elements) : m_data({elements...})
```

Comment:
No ambiguity with the previous constructor when `size() == 1` ?

Reply:
Hmmm, I added the other constructor later so I did not consider this. Indeed there could be an ambiguity here. But on the other hand a vector of length 1 is not super useful. I can't see why we would use a Vector in this case. But maybe it could arise in generic classes which work for multiple geometries